test/Chainweb/Test/Pact/PactSingleChainTest.hs

```diff
@@ -451,6 +457,72 @@ compactionIsIdempotent rdb =
                   putStrLn ""
       assertFailure "pact state check failed"
 
+compactionTransactionIndex :: ()
+  => RocksDb
+  -> TestTree
+compactionTransactionIndex rdb =
+  compactionSetup "compactionTransactionIndex" rdb testPactServiceConfig $ \cr -> do
+    setOneShotMempool cr.mempoolRef goldenMemPool
+
+    let makeTx :: Int -> BlockHeader -> IO ChainwebTransaction
+        makeTx nth bh = buildCwCmd testVersion
+          $ set cbSigners [mkEd25519Signer' sender00 [mkGasCap, mkTransferCap "sender00" "sender01" 1.0]]
+          $ setFromHeader bh
+          $ mkCmd (sshow (nth, bh))
+          $ mkExec' "(coin.transfer \"sender00\" \"sender01\" 1.0)"
+
+    supply <- newIORef @Int 0
+    madeTx <- newIORef @Bool False
+    let run = do
+          setMempool cr.mempoolRef $ mempty {
+            mpaGetBlock = \_ _ _ _ bh -> do
+              madeTxYet <- readIORef madeTx
+              if madeTxYet
+              then do
+                pure mempty
+              else do
+                n <- atomicModifyIORef' supply $ \a -> (a + 1, a)
+                tx <- makeTx n bh
+                writeIORef madeTx True
+                pure $ V.fromList [tx]
+          }
+          void $ runBlock cr.pactQueue cr.blockDb second
+          writeIORef madeTx False
+
+    let keepDepth :: Num a => a
+        keepDepth = 10
+
+    -- Run enough blocks to where TransactionIndex keeps its full history
+    replicateM_ keepDepth run
+
+    let db = _sConn cr.sqlEnv
+
+    let getRange :: IO (BlockHeight, BlockHeight)
+        getRange = do
+          r <- Pact.qry db "SELECT MIN(blockheight), MAX(blockheight) FROM TransactionIndex" [] [RInt, RInt]
+          case r of
+            [[SInt mn, SInt mx]] -> pure (int mn, int mx)
+            _ -> assertFailure "getRange: invalid query"
+
+    let compact = Utils.compact Error [C.NoVacuum, C.TransactionIndexKeepDepth keepDepth] cr.sqlEnv C.Latest
+
+    do
+      startRange <- getRange
+      compact
+      endRange <- getRange
+      assertEqual "TransactionIndex blockheight range is equal after compaction, when max blockheight is less than or equal to transactionIndexKeepDepth" startRange endRange
+
+    do
+      startRange <- getRange
+      let nMoreBlocks :: Num a => a
+          nMoreBlocks = 10
+      replicateM_ nMoreBlocks run
+      compact
+      endRange <- getRange
+      assertEqual "TransactionIndex blockheight lower range should shift after running more blocks and compacting" (fst startRange + nMoreBlocks) (fst endRange)
```

Comment:
these top two should be collapsed into one, eg `over both (+ nMoreBlocks) startRange == endRange`